jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -664,193 +175,103 @@ def mk_freeze_script(pub, locktime):
-    return tx
+    assert isinstance(tx, CMutableTransaction)
 
-def get_p2sh_signature(tx, i, redeem_script, pk, amount=None, hashcode=SIGHASH_ALL):
-    """
-    Tx is assumed to be serialized. redeem_script is for example the
-    output of mk_multisig_script.
-    pk is the private key, and must be passed in hex.
-    If amount is not None, the output of get_p2wsh_signature is returned.
-    What is returned is a single signature.
-    """
-    if isinstance(tx, str):
-        tx = binascii.unhexlify(tx)
-    if isinstance(redeem_script, str):
-        redeem_script = binascii.unhexlify(redeem_script)
-    if amount:
-        return get_p2wsh_signature(tx, i, redeem_script, pk, amount, hashcode)
-    modtx = signature_form(tx, i, redeem_script, hashcode)
-    return ecdsa_tx_sign(modtx, pk, hashcode)
+    pub = privkey_to_pubkey(priv)
 
-def get_p2wsh_signature(tx, i, redeem_script, pk, amount, hashcode=SIGHASH_ALL):
-    """ See note to multisign for the value to pass in as `script`.
-    Tx is assumed to be serialized.
-    """
-    modtx = segwit_signature_form(deserialize(tx), i, redeem_script, amount,
-                                  hashcode, decoder_func=lambda x: x)
-    return ecdsa_tx_sign(modtx, pk, hashcode)
+    if not amount:
+        # p2pkh only supported here:
+        input_scriptPubKey = pubkey_to_p2pkh_script(pub)
+        sighash = SignatureHash(input_scriptPubKey, tx, i, hashcode)
+        try:
+            sig = ecdsa_raw_sign(sighash, priv, rawmsg=True) + bytes([hashcode])
+        except Exception as e:
+            return return_err(e)
+        tx.vin[i].scriptSig = CScript([sig, pub])
+        # Verify the signature worked.
+        try:
+            VerifyScript(tx.vin[i].scriptSig,
+                        input_scriptPubKey, tx, i, flags=flags)
+        except Exception as e:
+            return return_err(e)
+        return sig, "signing succeeded"
 
-def apply_p2wsh_multisignatures(tx, i, script, sigs):
-    """Sigs must be passed in as a list, and must be a
-    complete list for this multisig, in the same order
-    as the list of pubkeys when creating the scriptPubKey.
-    """
-    if isinstance(script, str):
-        script = binascii.unhexlify(script)
-    sigs = [binascii.unhexlify(x) if x[:2] == '30' else x for x in sigs]
-    if isinstance(tx, str):
-        return safe_hexlify(apply_p2wsh_multisignatures(
-            binascii.unhexlify(tx), i, script, sigs))
-    txobj = deserialize(tx)
-    txobj["ins"][i]["script"] = ""
-    txobj["ins"][i]["txinwitness"] = [None] + sigs + [script]
-    return serialize(txobj)
+    else:
+        # segwit case; we currently support p2wpkh native or under p2sh.
 
-def apply_multisignatures(*args):
-    # tx,i,script,sigs OR tx,i,script,sig1,sig2...,sig[n]
-    tx, i, script = args[0], int(args[1]), args[2]
-    sigs = args[3] if isinstance(args[3], list) else list(args[3:])
+        # see line 1256 of bitcointx.core.scripteval.py:
```

Comment:
Concrete code line may become obsolete. If lined via line number, I think a tag or commit hash should be given. Or maybe just say 'p2sh flags are required for segwit'

Reply:
Yes, thanks for reminding me, I noticed this was dumb on my last review :) Addressed in https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/9933408ad70e717014ecbb24796d104f3263ac05